### ISSUE TITLE
Policy example: Evolving policies, two stage processing

### ISSUE THREAD

**Issue by seefeldb** [project-oak/oak#1412]
This is an example policy need, see #1410.

It explores changing policies for stored data, whether growing or shrinking in what they require/allow.

This is a more complex example along the lines of #1411.

---

As an example, let's assume two kinds of processors: One is `non_eu_redactor` which reads a coordinate and outputs the coordinate or null, depending on whether it is in the EU. And `at_most_city_redactor`, which reads a coordinate and outputs a coordinate rounded to the closest city of some minimal size.

In v1 of my application, coordinates are read and processed by `non_eu_redactor` , written into a database that retains labels. Then the entire database is read, aggregated by `aggregator` and published. The policy mandates that coordinates have to be within the EU and that those need to be aggregated before publishing.

In v2, coordinates are first processed by `non_eu_redactor` and then `at_most_city_redactor`, written into a database that retains labels. Then the entire database is read, aggregated by `aggregator` and published. The policy mandates that coordinates have to be within the EU_ and rounded to the next city_, and that those need to be aggregated before publishing (the _italic_ part is new in v2).

What happens to data that was written into the database under v1, but read under v2?

**Requirement 1: The data generated in v1 would not be published by v2, as in the meantime the policy got tighter.**

As an extension of the example, let’s weaken the policy:

In v3, the policy no longer requires aggregation of the data before publishing it. Nor does it require any redaction. Essentially the policy makes everything public. (Scenario: A malicious entity has taken over the app, or more precisely the principal the user elected as delegate to determine the policy).

**Requirement 2: v3 can update the policy, but can’t make it weaker than some floor that v1 and v2 specified, e.g. requiring some minimal amount of aggregation.**

_Before we get into detailed questions, is that example clear?_

**Comment by tiziano88** [comment]
Let me try and rephrase some of this in a different way, to see if it helps.

- Confidentiality lattice has L, M, H elements, in the obvious order (L ⊑ M ⊑ H).
- Input data is labelled as H.
- p0 has privilege to declassify data from H to M.
- p1 has no declassification privilege.
- Input data gets processed by p0, declassified from H to M, possibly stored somewhere with label M.
- p1 also processes the data, keeps label M.
- at some point in the future someone decides that data labelled as M can actually be treated as L.

The last one is the important step, and which IMO is being handwaved over, which, in the example above, corresponds to the statement "in v2 [...] the policies mandates [...]". What is such policy? Who can change it? How does it reach all the nodes?

Currently this is not modelled as an IFC concept, but we should see if it fits into that model.

It feels like it cannot be a label, since labels are constant once assigned, so we cannot just say "treat M as L", also because it is unclear who should have the authority to do so in the first place.

Perhaps it is a modification of the underlying lattice, in which the flow M ⊑ L is added at some point in time? Even in this case, it must be clarified who has the authority of doing so.

Basically, the crucial step is the jump from v1 to v2, but it is not described in the example above.

**Comment by tiziano88** [comment]
Reading more closely, it seems perhaps you are assuming that v1 and v2 **of the same app** are somewhat related? i.e. that Oak would keep track of versions between apps? But I note that's not the current model; in Oak each app is a set of nodes, and data stored by a different "version" of an app is available to any subsequent app, according to IFC labels of course. So perhaps this is an aspect we should explore to make sure we are on the same page.

**Comment by seefeldb** [comment]
I'm not assuming Oak would have to track version changes, this scenario should also work if two different apps happened to write into the same store.

I don't want to define how exactly policy changing works in the problem description (generally, avoiding solutions there), but yes it's a requirement that we add some way of evolving policies (and applications).

In this example p1 isn't running in v1, just p0. I'm not sure I follow the labels you propose, which one can be removed by the aggregation node?

I suppose a solution for the first section could be that we use a different aggregator in v2, and so no data from v1 can be released, because it has the wrong aggregator label. But then the protection is just from choosing not to run the old aggregator, but the policy iteration should disallow that. So upgrading the policy means revoking the old aggregator? Possible, but messy.

This can be solved quite directly with integrity labels of the derivation sort + policy upgrading as late binding translation of labels on the aggregation node. Will write that up later.
